### PR TITLE
ci: Consolidate requirements into `pyproject.toml`

### PR DESCRIPTION
**Pull request by CaedenPH**
### Describe your change:
Consolidate the requirements into `pyproject.toml`
https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/

* [ ] Add an algorithm?
* [x] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by CaedenPH** [comment]
@cclauss Never used pyproject.toml, so expecting lots of changes

**Comment by cclauss** [comment]
We could also move our codespell, mypy, and pyupgrade args from .pre-commit into pyproject.toml.

Please run https://pypi.org/project/validate-pyproject to validate the file and let's [add that tool to pre-commit](https://github.com/internetarchive/openlibrary/blob/master/.pre-commit-config.yaml#L57-L60).

**Comment by cclauss** [comment]
We might run into trouble consolidating our pytest args because our GitHub Actions run it [two](https://github.com/TheAlgorithms/Python/blob/master/.github/workflows/build.yml#L25) different [ways](https://github.com/TheAlgorithms/Python/blob/master/.github/workflows/project_euler.yml#L25).  So that might be difficult to express.

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,43 @@
+[project]
+name = "python-algorithms"
+description = "All algorithms implemented in Python"
+version = "1.0.0"
+
+authors = [
+    { name = "Christian Clauss", email = "cclauss@me.com" },
+    { name = "Dhruv Manilawala", email = "dhruvmanila@gmail.com" },
+    { name = "John law", email = "dhruvmanila@gmail.com" }
```
Comment (cclauss) [inline pyproject.toml:9]:
```suggestion
    { name = "John Law", email = "dhruvmanila@gmail.com" }
```

**Comment by CaedenPH** [comment]
> We could also move our codespell, mypy, and pyupgrade args from .pre-commit into pyproject.toml.
> 
> Please run https://pypi.org/project/validate-pyproject to validate the file and let's [add that tool to pre-commit](https://github.com/internetarchive/openlibrary/blob/master/.pre-commit-config.yaml#L57-L60).

How would this be done?

File: pyproject.toml [hunk 2]
```diff
@@ -6,7 +6,7 @@ version = "1.0.0"
 authors = [
     { name = "Christian Clauss", email = "cclauss@me.com" },
     { name = "Dhruv Manilawala", email = "dhruvmanila@gmail.com" },
-    { name = "John law", email = "dhruvmanila@gmail.com" }
+    { name = "John Law", email = "dhruvmanila@gmail.com" }
```
Comment (cclauss) [inline pyproject.toml:9]: The email address is wrong.

Comment (CaedenPH) [reply pyproject.toml:9]: > The email address is wrong.

What is the correct email address?
I took it from his github bio

Comment (cclauss) [reply pyproject.toml:9]: @poyea Do you want to be listed here?  With your email address?

**Comment by dhruvmanila** [comment]
The only benefit we get in using `pyproject.toml` is to have a central place for configuration. We don't need the project declaraction, build system, dependencies.

Build system is used to provide a backend like poetry, flit or setuptools to actually build the Python package (wheels). For our use case, `requirements.txt` is actually the best option instead of declaring it in `pyproject.toml`. Otherwise, users will be required to install using `pip install .` which installs the dependencies along with the *package* (remember this is not a package) and the egg info. Project declaration is not useful as it's used to provide metadata information for PyPi. I guess it was added to declare the dependencies which as mentioned not useful. Adding the authors field is a bit misleading as this is a work of all the contributors.

**Comment by cclauss** [comment]
Let's use `maintainers` rather than `authors`... https://peps.python.org/pep-0621/#authors-maintainers

Would we be able to use `pip install .` or probably better `pip install --editable .` in our GitHub Actions?

Could `pip install git+https://github.com/TheAlgorithms/Python.git` be made to work?

**Comment by dhruvmanila** [comment]
As I said, this won't work as `TheAlgorithms/Python` is not a "Python package", you cannot install it as it's not even structured like a standard package. I don't really see why you're wasting time in this.

**Comment by cclauss** [comment]
OK...  So now pip seems to be installing all of our dependencies and dev-dependencies.  It is pytest that seems broken.

**Comment by cclauss** [comment]
Now it seems to be working end-to-end without the need for `requirements.txt`!!

File: pyproject.toml (hunk 2 above)
Comment (poyea) [reply pyproject.toml:13]: I'm good with that

**Comment by cclauss** [comment]
Please rebase.

**Comment by CaedenPH** [comment]
> Please rebase.

Done

**Comment by dhruvmanila** [comment]
I don't really recommend defining dependencies in `pyproject.toml` file as this is not a Python package. It doesn't make sense to define them there. We can use `pyproject.toml` to define all the tool configs though, but we should be using `requirements.txt` to define the dependencies.

**Comment by cclauss** [comment]
I fail to understand the logic.  `pip` knows how to install project dependencies from two different file types.  Why does it matter whether this is a package or not?

**Review by dhruvmanila — CHANGES_REQUESTED** [review]
I guess I must've misunderstood something, if it works then it's fine @cclauss

File: pyproject.toml [hunk 3]
```diff
@@ -1,3 +1,52 @@
+[build-system]
+requires = ["setuptools ~= 65.5", "wheel"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "python-algorithms"
+description = "All algorithms implemented in Python"
+version = "1.0.0"
+
+authors = [
```
[the remark this answers is not in the document]
Comment (cclauss) [reply pyproject.toml:10]: https://peps.python.org/pep-0621#drop-maintainers-to-unify-with-authors

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -7,7 +7,10 @@ repos:
       - id: end-of-file-fixer
         types: [python]
       - id: trailing-whitespace
-      - id: requirements-txt-fixer
+        exclude: |
+          (?x)^(
+            data_structures/heap/binomial_heap.py
+          )$
```
Comment (dhruvmanila) [inline .pre-commit-config.yaml:13]: This got removed here: https://github.com/TheAlgorithms/Python/commit/57ccabbaeb0f32165271e3a218bc9c6dcfc21823. I don't know how it got added back here?

**Review by dhruvmanila — CHANGES_REQUESTED** [review]

**Review by dhruvmanila — APPROVED** [review]
Thanks a lot! 🎉

**Review by poyea — APPROVED** [review]
Thank you!

File: pyproject.toml [hunk 5]
```diff
@@ -1,3 +1,53 @@
...
+    "numpy",
+    "opencv-python",
+    "pandas",
+    "pillow",
+    "projectq",
+    "qiskit",
+    "requests",
+    "rich",
+    "scikit-fuzzy",
+    "sklearn",
```
Comment (cclauss) [inline pyproject.toml:31]:
```suggestion
    "scikit-learn",
```
https://pypi.org/project/scikit-learn/

File: pyproject.toml [hunk 6]
```diff
@@ -1,3 +1,53 @@
...
+    "statsmodels",
+    "sympy",
+    "tensorflow",
```
Comment (cclauss) [inline pyproject.toml:34]:
```suggestion
    # "tensorflow",  # Not yet compatible with Python 3.11
```

**Pull request closed without merging**